File: CSharp_Project/ParallelPLab4/ParallelPLab4/AsymmetricWay.cs
using System;
using System.Threading;

namespace ParallelPLab4
{
    public class TableAsymmetric
    {
        private readonly Semaphore[] forks = new Semaphore[5];

        public TableAsymmetric()
        {
            for (int i = 0; i < 5; i++)
            {
                forks[i] = new Semaphore(1, 1);
            }
        }
        public void GetFork(int forkId) => forks[forkId].WaitOne();
        public void PutFork(int forkId) => forks[forkId].Release();
    }

    public class PhilosopherAsymmetric
    {
        private readonly TableAsymmetric table;
        private readonly int firstFork, secondFork;
        private readonly int id;

        public PhilosopherAsymmetric(int id, TableAsymmetric table)
        {
            this.id = id;
            this.table = table;
            int leftFork = id;
            int rightFork = (id + 1) % 5;
            if (id == 4)
            {
                firstFork = rightFork;
                secondFork = leftFork;
            }
            else
            {
                firstFork = leftFork;
                secondFork = rightFork;
            }
        }

        public void Start()
        {
            for (int i = 0; i < 1; i++)
            {
                Console.WriteLine($"Philosopher {id} is thinking.");
                Thread.Sleep(500);

                table.GetFork(firstFork);
                Console.WriteLine($"Philosopher {id} picked up first fork {firstFork}");
                table.GetFork(secondFork);
                Console.WriteLine($"Philosopher {id} picked up second fork {secondFork}");
                Console.WriteLine($"Philosopher {id} is eatting");
                Thread.Sleep(1000);
                table.PutFork(secondFork);
                Console.WriteLine($"Philosopher {id} put down second fork {secondFork}");
                table.PutFork(firstFork);
                Console.WriteLine($"Philosopher {id} put down first fork {firstFork}");
            }
        }
    }
}

File: Kotlin_Project/ParallelPLab4/src/AsymmetricWay.kt
class PhilosopherAsymmetric(private val id: Int, private val table: Table) : Runnable {
    private val firstFork: Int
    private val secondFork: Int

    init {
        val leftFork = id
        val rightFork = (id + 1) % 5
        if (id == 4) {
            firstFork = rightFork
            secondFork = leftFork
        } else {
            firstFork = leftFork
            secondFork = rightFork
        }
    }

    override fun run() {
        try {
            println("Philosopher $id is thinking.")
            Thread.sleep(500)

            table.getFork(firstFork)
            println("Philosopher $id picked up first fork $firstFork")

            table.getFork(secondFork)
            println("Philosopher $id picked up second fork $secondFork")

            println("Philosopher $id is eating")
            Thread.sleep(1000)

            table.putFork(secondFork)
            println("Philosopher $id put down left fork $secondFork")

            table.putFork(firstFork)
            println("Philosopher $id put down right fork $secondFork")
        } catch (e: InterruptedException) {
            Thread.currentThread().interrupt()
        }
    }
}